src/rust/engine/process_execution/src/docker.rs

```diff
@@ -295,222 +352,331 @@ impl super::CommandRunner for CommandRunner {
-          .status_code;
+    Ok(container.id)
+  }
 
-        break status_code;
-      };
+  /// Return the container ID of a container running `image` for use as a place to invoke
+  /// build actions as executions within the cached container.
+  pub async fn container_id_for_image(
+    &self,
+    image: &str,
+    build_generation: &str,
+  ) -> Result<String, String> {
+    let docker = self.docker.get().await?;
+    let docker = docker.clone();
 
-      log::trace!("container {} exited with status code {}", &container_id, status_code);
+    let container_id_cell = {
+      let mut containers = self.containers.lock();
+      let cell = containers
+        .entry(image.to_string())
+        .or_insert_with(|| Arc::new(OnceCell::new()));
+      cell.clone()
+    };
 
-      let attach_options = bollard::container::AttachContainerOptions::<String> {
-        stdout: Some(true),
-        stderr: Some(true),
-        logs: Some(true),
-        ..bollard::container::AttachContainerOptions::default()
-      };
+    let work_dir_base = self.work_dir_base.clone();
+    let named_caches_base_dir = self.named_caches_base_dir.clone();
+    let immutable_inputs_base_dir = self.immutable_inputs_base_dir.clone();
+    let image_pull_cache = Arc::clone(&self.image_pull_cache);
 
-      let attach_result = docker
-        .attach_container(&container.id, Some(attach_options))
+    let container_id = container_id_cell
+      .get_or_try_init(async move {
+        Self::make_container(
+          docker,
+          image_pull_cache,
+          image.to_string(),
+          build_generation,
+          work_dir_base,
+          named_caches_base_dir,
+          immutable_inputs_base_dir,
+        )
         .await
-        .map_err(|err| {
-          format!(
-            "Failed to attach to Docker container `{}`: {:?}",
-            &container_id, err
-          )
-        })?;
-
-      log::trace!("attached to container {}", &container.id);
+      })
+      .await?;
 
-      let mut output_stream = attach_result.output.boxed();
+    Ok(container_id.to_owned())
+  }
+}
 
-      while let Some(output_msg) = output_stream.next().await {
-        match output_msg {
-            Ok(LogOutput::StdOut { message }) => {
-                log::trace!("container {} wrote {} bytes to stdout", &container_id, message.len());
-                yield ChildOutput::Stdout(message);
-            }
-            Ok(LogOutput::StdErr { message }) => {
-                log::trace!("container {} wrote {} bytes to stderr", &container_id, message.len());
-                yield ChildOutput::Stderr(message);
-            }
-            Ok(_) => (),
-            Err(err) => {
-                log::trace!("error while capturing output of container {}: {:?}", &container_id, err);
-            }
+impl Drop for ContainerCache {
+  fn drop(&mut self) {
+    let docker = self.docker.clone();
+    let container_ids = self.containers.lock().keys().cloned().collect::<Vec<_>>();
```

Comment:
This codepath probably isn't performance-sensitive enough for it to be worth it, but:
```suggestion
    let containers = std::mem::take(&mut self.containers.lock());
```

Reply:
Didn't work as the take was seen as taking the MutexGuard, not the cotents.

Reply:
Usually that means you need some combination of ref/deref'ing... maybe:
```rust
std::mem::take(&mut (*self.containers.lock()))
```
But yea, not a big deal.